File: src/fuser/memory_base/indexer.py
import hashlib
import logging
from datetime import datetime, timedelta
from pathlib import Path

import numpy as np

from fuser.knowledge_base.base_embedding import BaseEmbeddingClient
from fuser.knowledge_base.base_retriever import Document

DEFAULT_MIN_SCORE = 0.3
DEFAULT_VALID_DURATION_DAYS = 14


def _hash_text(text: str) -> str:
    """SHA-256 hash of text content."""
    return hashlib.sha256(text.encode("utf-8")).hexdigest()


def _cosine_similarity(a: np.ndarray, b: np.ndarray) -> float:
    """Cosine similarity between two vectors."""
    norm_a = np.linalg.norm(a)
    norm_b = np.linalg.norm(b)
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return float(np.dot(a, b) / (norm_a * norm_b))


class MemoryIndex:
    """In-memory embedding index for memory chunks.

    Parameters
    ----------
    embedding_client : BaseEmbeddingClient
        Embedding client from knowledge base
    """

    def __init__(self, embedding_client: BaseEmbeddingClient):
        self.embedding_client = embedding_client
        self._cache: dict[str, tuple[np.ndarray, Document]] = {}

    @property
    def size(self) -> int:
        """Number of cached chunks."""
        return len(self._cache)

    async def search(self, query: str, top_k: int = 1, min_score: float = DEFAULT_MIN_SCORE) -> list[Document]:
        """Search for most similar chunks using cosine similarity.

        Parameters
        ----------
        query : str
            The search query text.
        top_k : int
            Number of top results to return.
        min_score : float
            Minimum similarity score threshold.

        Returns
        -------
        list of Document
            Top-k matching documents with similarity scores.
        """
        if not self._cache or not query.strip():
            return []

        try:
            async with self.embedding_client:
                query_embedding = await self.embedding_client.embed(query)
        except Exception as e:
            logging.error(f"Memory index: failed to embed query: {e}")
            return []

        scored: list[tuple[float, Document]] = []
        for embedding, doc in self._cache.values():
            score = _cosine_similarity(query_embedding, embedding)
            if score >= min_score:
                scored.append((score, doc))

        scored.sort(key=lambda x: x[0], reverse=True)
        results = []
        for score, doc in scored[:top_k]:
            results.append(Document(text=doc.text, metadata=doc.metadata.copy(), score=score))

        logging.info(f"Memory search: '{query[:50]}' → {len(results)} results " f"(from {self.size} chunks)")
        return results

    async def load_chunks_batch(self, chunks: list[Document]) -> int:
        """Load multiple chunks into the index in batch (used at startup).

        Parameters
        ----------
        chunks : list of Document
            Chunks to load.

        Returns
        -------
        int
            Number of new chunks actually embedded.
        """
        new_chunks = []
        new_hashes = []
        for chunk in chunks:
            text_hash = _hash_text(chunk.text)
            if text_hash not in self._cache:
                new_chunks.append(chunk)
                new_hashes.append(text_hash)

        if not new_chunks:
            return 0

        try:
            texts = [c.text for c in new_chunks]
            async with self.embedding_client:
                embeddings = await self.embedding_client.embed_batch(texts)

            for i, (text_hash, chunk) in enumerate(zip(new_hashes, new_chunks)):
                self._cache[text_hash] = (embeddings[i], chunk)

            logging.info(f"Memory index: loaded {len(new_chunks)} chunks (total: {self.size})")
            return len(new_chunks)
        except Exception as e:
            logging.error(f"Memory index: batch embed failed: {e}")
            return 0


def parse_daily_file(filepath: Path) -> list[Document]:
    """Parse a daily markdown file into document chunks.

    Each ## section becomes a separate chunk.

    Parameters
    ----------
    filepath : Path
        Path to a daily log markdown file.

    Returns
    -------
    list of Document
        Parsed document chunks with metadata.
    """
    try:
        content = filepath.read_text(encoding="utf-8")
    except Exception as e:
        logging.error(f"Memory: failed to read {filepath}: {e}")
        return []

    date_prefix = f"[Date: {filepath.stem}]"

    chunks: list[Document] = []
    current_chunk = ""
    chunk_start_line = 1

    for i, line in enumerate(content.split("\n"), 1):
        if line.startswith("## ") and current_chunk.strip():
            chunks.append(
                Document(
                    text=f"{date_prefix}\n{current_chunk.strip()}",
                    metadata={
                        "source": filepath.name,
                        "chunk_id": len(chunks),
                        "start_line": chunk_start_line,
                    },
                )
            )
            current_chunk = line + "\n"
            chunk_start_line = i
        else:
            current_chunk += line + "\n"

    if current_chunk.strip():
        chunks.append(
            Document(
                text=f"{date_prefix}\n{current_chunk.strip()}",
                metadata={
                    "source": filepath.name,
                    "chunk_id": len(chunks),
                    "start_line": chunk_start_line,
                },
            )
        )

    return chunks


async def build_index(
    index: MemoryIndex,
    daily_dir: Path,
    valid_duration: int = DEFAULT_VALID_DURATION_DAYS,
) -> None:
    """Build an existing MemoryIndex from recent daily markdown files.

    Expired files (older than ``valid_duration`` days) are deleted.

    Parameters
    ----------
    index : MemoryIndex
        The index instance to populate (mutated in place).
    daily_dir : Path
        Directory containing daily markdown files.
    valid_duration : int
        Number of days to keep. Older daily logs are deleted.
    """
    if not daily_dir.exists():
        return

    cutoff = datetime.now() - timedelta(days=valid_duration)
    all_chunks: list[Document] = []

    for daily_file in sorted(daily_dir.glob("*.md")):
        try:
            file_date = datetime.strptime(daily_file.stem, "%Y-%m-%d")
        except ValueError:
            continue

        if file_date < cutoff:
            daily_file.unlink()
            logging.info(f"Memory: deleted expired daily log {daily_file.name}")
            continue

        chunks = parse_daily_file(daily_file)
        all_chunks.extend(chunks)

    if all_chunks:
        count = await index.load_chunks_batch(all_chunks)
        logging.info(
            f"Memory: populated index with {len(all_chunks)} chunks, {count} new "
            f"(valid_duration: {valid_duration} days)"
        )

File: internal/memory/indexer.go
package memory

import (
	"bufio"
	"context"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/coder/hnsw"
	"go.uber.org/zap"

	"github.com/openmind/om1/internal/knowledgebase"
)

const (
	DefaultMinScore          = 0.5
	DefaultValidDurationDays = 60
)

// MemoryIndex is an in-memory HNSW index for memory chunks.
type MemoryIndex struct {
	mu       sync.RWMutex
	graph    *hnsw.Graph[int]
	docs     map[int]MemoryEntry
	hashes   map[string]int
	nextID   int
	bm25     *BM25Index
	embedder knowledgebase.Embedder
	log      *zap.Logger
}

// NewMemoryIndex creates a new index backed by the given embedder.
func NewMemoryIndex(embedder knowledgebase.Embedder, log *zap.Logger) *MemoryIndex {
	return &MemoryIndex{
		graph:    hnsw.NewGraph[int](),
		docs:     make(map[int]MemoryEntry),
		hashes:   make(map[string]int),
		bm25:     NewBM25Index(),
		embedder: embedder,
		log:      log,
	}
}

// Size returns the number of indexed chunks.
func (idx *MemoryIndex) Size() int {
	idx.mu.RLock()
	defer idx.mu.RUnlock()
	return len(idx.docs)
}

// Search finds the top-k most similar chunks to query using HNSW ANN search.
func (idx *MemoryIndex) Search(ctx context.Context, query string, topK int, minScore float64, userID string) ([]MemoryEntry, error) {
	idx.mu.RLock()
	totalChunks := len(idx.docs)
	idx.mu.RUnlock()

	if totalChunks == 0 || strings.TrimSpace(query) == "" {
		return nil, nil
	}

	queryVec, err := idx.embedder.Embed(ctx, query)
	if err != nil {
		return nil, fmt.Errorf("memory index: embed query: %w", err)
	}

	idx.mu.RLock()
	defer idx.mu.RUnlock()

	searchK := topK * 3
	if searchK > len(idx.docs) {
		searchK = len(idx.docs)
	}

	nodes := idx.graph.Search(queryVec, searchK)

	var results []MemoryEntry
	for _, node := range nodes {
		doc, ok := idx.docs[node.Key]
		if !ok {
			continue
		}
		similarity := 1 - float64(idx.graph.Distance(queryVec, node.Value))
		if similarity >= minScore {
			doc.Score = similarity
			results = append(results, doc)
		}
		if len(results) >= topK {
			break
		}
	}

	idx.log.Info("memory search",
		zap.String("query", truncate(query, 50)),
		zap.Int("results", len(results)),
		zap.Int("total_chunks", totalChunks),
	)
	return results, nil
}

const (
	fusionAlpha = 0.7 // embedding weight
	fusionBeta  = 0.3 // BM25 weight
	userBoost   = 0.1 // soft boost for matching user_id
)

// HybridSearch combines embedding cosine similarity with BM25 lexical scoring.
func (idx *MemoryIndex) HybridSearch(ctx context.Context, query string, topK int, minScore float64, userID string) ([]MemoryEntry, error) {
	embeddingResults, err := idx.Search(ctx, query, topK*2, minScore, userID)
	if err != nil {
		return nil, err
	}

	bm25Results := idx.bm25.Search(query, topK*2, userID)

	// Build score maps keyed by text hash.
	type fusedEntry struct {
		doc            MemoryEntry
		embeddingScore float64
		bm25Score      float64
	}
	merged := make(map[string]*fusedEntry)

	for _, r := range embeddingResults {
		h := hashText(r.Text)
		merged[h] = &fusedEntry{doc: r, embeddingScore: r.Score}
	}
	for _, r := range bm25Results {
		h := hashText(r.Text)
		if e, ok := merged[h]; ok {
			e.bm25Score = r.Score
		} else {
			merged[h] = &fusedEntry{doc: r, bm25Score: r.Score}
		}
	}

	if len(merged) == 0 {
		return nil, nil
	}

	// Min-max normalize each signal.
	var eMin, eMax, bMin, bMax float64
	first := true
	for _, e := range merged {
		if first {
			eMin, eMax = e.embeddingScore, e.embeddingScore
			bMin, bMax = e.bm25Score, e.bm25Score
			first = false
		}
		if e.embeddingScore < eMin {
			eMin = e.embeddingScore
		}
		if e.embeddingScore > eMax {
			eMax = e.embeddingScore
		}
		if e.bm25Score < bMin {
			bMin = e.bm25Score
		}
		if e.bm25Score > bMax {
			bMax = e.bm25Score
		}
	}

	normalize := func(v, min, max float64) float64 {
		if max == min {
			return 0
		}
		return (v - min) / (max - min)
	}

	type scored struct {
		doc   MemoryEntry
		score float64
	}
	var results []scored
	for _, e := range merged {
		final := fusionAlpha*normalize(e.embeddingScore, eMin, eMax) +
			fusionBeta*normalize(e.bm25Score, bMin, bMax)

		if userID != "" && e.doc.Metadata["user_id"] == userID {
			final += userBoost
		}

		d := e.doc
		d.Score = final
		results = append(results, scored{doc: d, score: final})
	}

	// Sort descending.
	for i := 1; i < len(results); i++ {
		for j := i; j > 0 && results[j].score > results[j-1].score; j-- {
			results[j], results[j-1] = results[j-1], results[j]
		}
	}
	if len(results) > topK {
		results = results[:topK]
	}

	docs := make([]MemoryEntry, len(results))
	for i, r := range results {
		docs[i] = r.doc
	}

	idx.log.Info("memory hybrid search",
		zap.String("query", truncate(query, 50)),
		zap.Int("embedding_hits", len(embeddingResults)),
		zap.Int("bm25_hits", len(bm25Results)),
		zap.Int("fused", len(docs)),
	)
	return docs, nil
}

// LoadChunksBatch loads multiple chunks into the index in one batch on startup.
func (idx *MemoryIndex) LoadChunksBatch(ctx context.Context, chunks []MemoryEntry) (int, error) {
	idx.mu.RLock()
	var newChunks []MemoryEntry
	var newHashes []string
	for _, c := range chunks {
		h := hashText(c.Text)
		if _, exists := idx.hashes[h]; !exists {
			newChunks = append(newChunks, c)
			newHashes = append(newHashes, h)
		}
	}
	idx.mu.RUnlock()

	if len(newChunks) == 0 {
		return 0, nil
	}

	idx.mu.Lock()
	defer idx.mu.Unlock()

	loaded := 0
	for i, chunk := range newChunks {
		vec, err := idx.embedder.Embed(ctx, chunk.Text)
		if err != nil {
			idx.log.Warn("memory index: batch embed failed, partial load",
				zap.Int("loaded", loaded), zap.Error(err))
			return loaded, err
		}
		id := idx.nextID
		idx.nextID++
		idx.graph.Add(hnsw.MakeNode(id, vec))
		idx.docs[id] = chunk
		idx.hashes[newHashes[i]] = id
		idx.bm25.Add(newHashes[i], chunk)
		loaded++
	}

	idx.log.Info("memory index: loaded chunks",
		zap.Int("new", loaded),
		zap.Int("total", len(idx.docs)),
	)
	return loaded, nil
}

// AddChunk embeds and inserts a single chunk (hot update). Returns true if added.
func (idx *MemoryIndex) AddChunk(ctx context.Context, chunk MemoryEntry) (bool, error) {
	h := hashText(chunk.Text)

	idx.mu.RLock()
	_, exists := idx.hashes[h]
	idx.mu.RUnlock()
	if exists {
		return false, nil
	}

	vec, err := idx.embedder.Embed(ctx, chunk.Text)
	if err != nil {
		return false, fmt.Errorf("memory index: hot update: %w", err)
	}

	idx.mu.Lock()
	id := idx.nextID
	idx.nextID++
	idx.graph.Add(hnsw.MakeNode(id, vec))
	idx.docs[id] = chunk
	idx.hashes[h] = id
	idx.mu.Unlock()

	idx.bm25.Add(h, chunk)

	idx.log.Debug("memory index: hot updated", zap.Int("total", idx.Size()))
	return true, nil
}

type persistedEntry struct {
	ID       int               `json:"id"`
	Hash     string            `json:"hash"`
	Text     string            `json:"text"`
	Metadata map[string]string `json:"metadata"`
}

type persistedIndex struct {
	NextID  int              `json:"next_id"`
	Entries []persistedEntry `json:"entries"`
}

// SaveToDisk persists the HNSW graph and document metadata to dir.
func (idx *MemoryIndex) SaveToDisk(dir string) error {
	idx.mu.RLock()
	defer idx.mu.RUnlock()

	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("memory index: create dir: %w", err)
	}

	// Save HNSW graph.
	graphPath := filepath.Join(dir, "index.graph")
	f, err := os.Create(graphPath)
	if err != nil {
		return fmt.Errorf("memory index: create graph file: %w", err)
	}
	defer func() { _ = f.Close() }()
	w := bufio.NewWriter(f)
	if err := idx.graph.Export(w); err != nil {
		return fmt.Errorf("memory index: export graph: %w", err)
	}
	if err := w.Flush(); err != nil {
		return fmt.Errorf("memory index: flush graph: %w", err)
	}

	// Save document metadata.
	idToHash := make(map[int]string, len(idx.hashes))
	for h, id := range idx.hashes {
		idToHash[id] = h
	}
	entries := make([]persistedEntry, 0, len(idx.docs))
	for id, doc := range idx.docs {
		entries = append(entries, persistedEntry{
			ID: id, Hash: idToHash[id], Text: doc.Text, Metadata: doc.Metadata,
		})
	}
	data, err := json.Marshal(persistedIndex{NextID: idx.nextID, Entries: entries})
	if err != nil {
		return fmt.Errorf("memory index: marshal metadata: %w", err)
	}
	if err := os.WriteFile(filepath.Join(dir, "index.meta.json"), data, 0o644); err != nil {
		return fmt.Errorf("memory index: write metadata: %w", err)
	}

	idx.log.Info("memory index: saved to disk", zap.Int("chunks", len(idx.docs)))
	return nil
}

// LoadFromDisk restores the HNSW graph and document metadata from dir.
func (idx *MemoryIndex) LoadFromDisk(dir string) error {
	idx.mu.Lock()
	defer idx.mu.Unlock()

	// Load HNSW graph.
	graphPath := filepath.Join(dir, "index.graph")
	f, err := os.Open(graphPath)
	if err != nil {
		return err
	}
	graph := hnsw.NewGraph[int]()
	if err := graph.Import(bufio.NewReader(f)); err != nil {
		_ = f.Close()
		return fmt.Errorf("memory index: import graph: %w", err)
	}
	_ = f.Close()

	// Load document metadata.
	raw, err := os.ReadFile(filepath.Join(dir, "index.meta.json"))
	if err != nil {
		return fmt.Errorf("memory index: read metadata: %w", err)
	}
	var meta persistedIndex
	if err := json.Unmarshal(raw, &meta); err != nil {
		return fmt.Errorf("memory index: parse metadata: %w", err)
	}

	idx.graph = graph
	idx.nextID = meta.NextID
	for _, e := range meta.Entries {
		doc := MemoryEntry{Text: e.Text, Metadata: e.Metadata}
		idx.docs[e.ID] = doc
		idx.hashes[e.Hash] = e.ID
		idx.bm25.Add(e.Hash, doc)
	}

	idx.log.Info("memory index: loaded from disk", zap.Int("chunks", len(idx.docs)))
	return nil
}

// PruneHashes removes entries whose hash is not in validHashes.
func (idx *MemoryIndex) PruneHashes(validHashes map[string]struct{}) int {
	idx.mu.Lock()
	defer idx.mu.Unlock()

	pruned := 0
	for h, id := range idx.hashes {
		if _, ok := validHashes[h]; !ok {
			delete(idx.docs, id)
			delete(idx.hashes, h)
			pruned++
		}
	}
	return pruned
}

// ParseDailyFile parses a daily markdown file into MemoryEntry chunks.
func ParseDailyFile(path string) ([]MemoryEntry, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("memory: read %s: %w", path, err)
	}

	base := filepath.Base(path)
	dateStem := strings.TrimSuffix(base, filepath.Ext(base))
	datePrefix := "[Date: " + dateStem + "]"

	userTagRe := regexp.MustCompile(`^\[User: (.+)\]$`)

	var chunks []MemoryEntry
	var currentChunk strings.Builder
	var currentUserID string

	lines := strings.Split(string(content), "\n")
	for _, line := range lines {
		if strings.HasPrefix(line, "## ") && currentChunk.Len() > 0 {
			text := strings.TrimSpace(currentChunk.String())
			if text != "" {
				meta := map[string]string{"source": base}
				if currentUserID != "" {
					meta["user_id"] = currentUserID
				}
				chunks = append(chunks, MemoryEntry{
					Text:     datePrefix + "\n" + text,
					Metadata: meta,
				})
			}
			currentChunk.Reset()
			currentUserID = ""
		}

		if m := userTagRe.FindStringSubmatch(line); m != nil {
			currentUserID = strings.ToLower(strings.TrimSpace(m[1]))
		}
		currentChunk.WriteString(line)
		currentChunk.WriteByte('\n')
	}

	// Flush last chunk.
	if text := strings.TrimSpace(currentChunk.String()); text != "" {
		meta := map[string]string{"source": base}
		if currentUserID != "" {
			meta["user_id"] = currentUserID
		}
		chunks = append(chunks, MemoryEntry{
			Text:     datePrefix + "\n" + text,
			Metadata: meta,
		})
	}

	return chunks, nil
}

// BuildIndex populates the given MemoryIndex from recent daily markdown files.
func BuildIndex(ctx context.Context, idx *MemoryIndex, dailyDir string, validDays int) error {
	entries, err := os.ReadDir(dailyDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}

	cutoff := time.Now().AddDate(0, 0, -validDays)
	var allChunks []MemoryEntry

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		stem := strings.TrimSuffix(entry.Name(), ".md")
		fileDate, parseErr := time.Parse("2006-01-02", stem)
		if parseErr != nil {
			continue
		}

		filePath := filepath.Join(dailyDir, entry.Name())
		if fileDate.Before(cutoff) {
			_ = os.Remove(filePath)
			idx.log.Info("deleted expired daily log", zap.String("file", entry.Name()))
			continue
		}

		chunks, parseErr := ParseDailyFile(filePath)
		if parseErr != nil {
			idx.log.Warn("failed to parse daily file", zap.String("file", entry.Name()), zap.Error(parseErr))
			continue
		}
		allChunks = append(allChunks, chunks...)
	}

	if len(allChunks) > 0 || idx.Size() > 0 {
		validHashes := make(map[string]struct{}, len(allChunks))
		for _, c := range allChunks {
			validHashes[hashText(c.Text)] = struct{}{}
		}
		if pruned := idx.PruneHashes(validHashes); pruned > 0 {
			idx.log.Info("pruned stale entries", zap.Int("count", pruned))
		}
	}

	if len(allChunks) > 0 {
		loaded, loadErr := idx.LoadChunksBatch(ctx, allChunks)
		if loadErr != nil {
			return loadErr
		}
		idx.log.Info("populated index",
			zap.Int("chunks", len(allChunks)),
			zap.Int("new", loaded),
			zap.Int("valid_days", validDays),
		)
	}
	return nil
}

func hashText(text string) string {
	h := sha256.Sum256([]byte(text))
	return fmt.Sprintf("%x", h)
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
